### ISSUE TITLE
Leadership Domains for Timelock Server

### ISSUE THREAD

**Issue by jeremyk-91** [palantir/atlasdb#1414]
This was an idea discussed with @rhero, @tboam and @harrybiddle.

One issue that can have a negative impact on Timelock Server perf is that we currently give a single leader the responsibility of handling lock and timestamp requests for all namespaces, while the rest of the boxes in the cluster do nothing.

It could be interesting to explore the idea of using leadership to split the responsibility across nodes. 

The simplest way could be to run multiple leadership election protocols in parallel, though this might probabilistically determine things and also require more heartbeats - more network traffic. Alternatively we can have the Paxos leader set out a schedule of responsibility and have other nodes take charge of their domain of responsibility as long as they can talk to the leader and the leader still confirms they "own" the relevant namespace. The key is that at any time we want there to be at most one node owning each namespace, but there's no need for this to be the same node (in fact, it's probably better for perf if they're distributed more evenly).

**Comment by tpetracca** [comment]
My understanding is that the majority of the performance bottlenecks in handing out timestamps is the consensus algorithm overhead, and more specifically the number of network calls between hosts to reach consensus.

If we're not bound in any sort of way on the actual CPU power of an individual machine what good does distributing who owns what namespaces? We'll still have to do all the same amount of network hops right?

**Comment by jeremyk-91** [comment]
I haven't done any profiling, though I was a little concerned about the network I/O overhead in handling a high volume of client requests, actually. You're probably right on the consensus being the main bottleneck though, given that for every client request we make N-1 inter-node calls since we try to talk to everyone in obtaining a quorum.

You're right that if there isn't a bottleneck at any individual machine then distributing isn't going to be effective - since indeed the number of network hops is unchanged (it might even increase).

This one isn't a high priority, in any case. (We don't plan on doing this until and unless we have perf issues, and even then I'd prefer to explore other approaches e.g. #1412 first.)

**Comment by jeremyk-91** [comment]
I'm reviving this and putting it in the new-issues queue again, since it does appear that to some extent individual TimeLock machines are falling over in our latest tests (after switching to HTTP/2).

Since the configuration is relatively static, a simple consistent hash could probably work. I think the number of clients is small and load balancing could be pretty useful, so I'm thinking [maglev hashing](https://github.com/kkdai/maglev), actually.

**Comment by gsheasby** [comment]
We should first investigate why leaders die so frequently - @harrybiddle will file a new issue that'll block this.

**Comment by harrybiddle** [comment]
See #1680 for investigation into frequent leadership changes.

**Comment by rhero** [comment]
If we do this, we should consider reopening #1712.

**Comment by gsheasby** [comment]
@jeremyk-91 given our recent investigations into MRTSE #1749, is this idea something we are likely to be pursuing soon?

**Comment by jeremyk-91** [comment]
Re-bumping to New Issues as this is expected to be a piece of the perf work I'll be picking up this week or next.

**Comment by hsaraogi** [comment]
Tracked via Graphlock issues.
